android: Update packaging to work with latest SDK

cnorthrop opened the pull request:
This changes our Android build process to call directly into the tools that create it, rather than use ant and "android project" which have been removed from the SDK.

The end result is the same APK, just signed differently, but equivalently.

Contents before:

```
$ aapt list ./bin/RenderDocCmd.apk
AndroidManifest.xml
res/drawable/icon.png
resources.arsc
classes.dex
lib/armeabi-v7a/librenderdoccmd.so
lib/armeabi-v7a/libVkLayer_GLES_RenderDoc.so
META-INF/MANIFEST.MF
META-INF/CERT.SF
META-INF/CERT.RSA
```

Contents after:

```
$ aapt list ./bin/RenderDocCmd.apk 
AndroidManifest.xml
classes.dex
lib/armeabi-v7a/libVkLayer_GLES_RenderDoc.so
lib/armeabi-v7a/librenderdoccmd.so
res/drawable/icon.png
resources.arsc
META-INF/ANDROIDD.SF
META-INF/ANDROIDD.RSA
META-INF/MANIFEST.MF
```

With this, developers can safely update their Android SDK components to latest.

Most of the tools in use come from the Android SDK, with the exception of **keytool**, which may need to be installed and added to PATH.

The following two env vars are needed and CMake will error out cleanly if they are not found: 
- **ANDROID_SDK**
- **JAVA_HOME**

libVkLayer_GLES_RenderDoc.so resides in the same place, so RenderDoc can continue to find it.

This also generates a debug.keystore that we can use with patching process, potentially ship.

Tested on Ubuntu and macOS.

cnorthrop commented:
I'll need to update the Travis-CI scripts on this branch as well.

baldurk commented:
This looks OK to me but I don't know much about android so waiting for @michaelrgb to comment.

FYI I won't merge it this week as I'm stabilising for release.

michaelrgb commented:
After adding the two new environment variables, I get 
`/bin/sh: 1: dx: not found`

Maybe that needs to be a full path based on the variables?

Also, perhaps the comment could list which deps are no longer required (such as ant).

cnorthrop commented:
I meant `$ANDROID_SDK`, not `$ANDROID_HOME`.  I've updated the branch to use it and `$JAVA_HOME` to find all the new dependencies, retested locally.

michaelrgb commented:
Looks good so far, but I decided to apt-get install android-sdk, which on 17.04 has version 24.0.0 of android-sdk-build-tools. Perhaps that hard-coded version number "25.0.2" could go in a variable?

I also had to apt-get install apksigner separately, as I dont think it exists inside my version of build-tools.

cnorthrop commented:
I would lean toward having a fixed build-tools version we support and have that baked in so we can always reproduce a build from a given hash.

Looks like the earliest build-tools that includes `apksigner` is `24.0.3`.  I'm not sure how we got on 25.0.*, maybe that was the latest when support was added.  I think if we're going to change versions, maybe roll forward to the latest public for everything?

I haven't used Debian packages to install the tools before, so I don't know off the top of my head how to get specific versions.  I always use either Android Studio to get them (you just have to check the boxes of the build-tools versions you want) or direct downloads via http as we do with Travis-CI.

If you have the `$ANDROID_SDK/tools` directory, you can use `sdkmanager` to get the version you need.  I'll add another commit to this branch that updates `CONTRIBUTING.md` for Android showing how to use it.

michaelrgb commented:
I agree bake/set it to the latest downloaded version (25.0.), but we can still just set it one place in a variable so that other people can use Debian if they want to.

Otherwise, I approve this change.

I find installing from Debian packages the most convenient way of getting the tools, but I accept most people want to download them instead.

cnorthrop commented:
Okay, I added an option to specify the build-tools version: `-DANDROID_BUILD_TOOLS_VERSION=foo`

I also added some Android documentation to CONTRIBUTING.md, since I was already putting together the requirements for Travis-CI.

cnorthrop commented:
I noticed this was failing the Travis-CI build.  I happened to have created the PR when master was briefly failing to build, regardless of the unique commits here.

I rebased this to the latest master, confirmed it builds locally.  Let me know if I should resubmit the PR against v1.x, which is not a problem.

baldurk commented on CONTRIBUTING.md:
````diff
@@ -40,6 +40,16 @@ Configuration is available for cmake, [documented elsewhere](https://cmake.org/d
+cd build-android
+cmake -DBUILD_ANDROID=On -DANDROID_ABI=armeabi-v7a -DANDROID_NATIVE_API_LEVEL=23 ..
+make
+```
````
Can you add a note here for windows builders?

Something like:

> On windows you need to specify the 'generator' type to the cmake invocation. The exact parameter will depend on your bash shell, but options are e.g. `-G "MSYS Makefiles"` or `-G "MinGW Makefiles"`

cnorthrop replied:
Okay, done.  I don't have the Windows for Android build set up, so I haven't verified the steps.  I tend to use cygwin for tracking projects and Visual Studio Command Prompt for command line compiles, including for Android.  I haven't tried getting that to work for RenderDoc - wasn't sure if anyone was actually compiling on Windows.

baldurk replied:
Thanks. I don't know about anyone else, but at least I'm building it on windows for the nightly builds. I imagine there are plenty of windows devs targeting android just because there are so many windows devs in general.

I think it's fine to expect the android subset of users to have bash available though on windows, since cmd is so awful.

cnorthrop replied:
Yes, awful, but easy to get going if you've got Visual Studio installed.  Compiling on Windows is important, so I'll get mingw going to ensure the steps always work.

baldurk commented on CONTRIBUTING.md:
````diff
@@ -149,6 +159,42 @@ brew install cmake qt5
+unzip android-ndk-r14b-linux-x86_64.zip
+export ANDROID_NDK=$ANDROID_SDK/android-ndk-r14b
+```
+
````
Can you also add instructions for fetching the dependencies on windows here? I think it should be roughly similar to the linux steps (and it's fine to just give wget/unzip commands for the relevant bits where possible) but since it's extra fiddly on windows it would be useful to have it listed here for reference.

cnorthrop replied:
I added instructions for Windows (and macOS while I was at it).

baldurk commented:
I tried this locally and I ran into some java errors during the build:

```
src\org\renderdoc\renderdoccmd\Loader.java:2: error: package android.app does not exist
import android.app.Activity;
                  ^
src\org\renderdoc\renderdoccmd\Loader.java:4: error: package android.app does not exist
public class Loader extends android.app.NativeActivity
                                       ^
src\org\renderdoc\renderdoccmd\Loader.java:12: error: package android.os does not exist
    protected void onCreate(android.os.Bundle savedInstanceState) {
                                      ^
src\org\renderdoc\renderdoccmd\Loader.java:11: error: method does not override or implement a method from a supertype
    @Override
    ^
src\org\renderdoc\renderdoccmd\Loader.java:13: error: cannot find symbol
        super.onCreate(savedInstanceState);
        ^
  symbol:   variable super
  location: class Loader
src\org\renderdoc\renderdoccmd\Loader.java:15: error: package android.Manifest does not exist
        requestPermissions(new String[]{android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
                                                        ^
src\org\renderdoc\renderdoccmd\Loader.java:15: error: cannot find symbol
        requestPermissions(new String[]{android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, 1);
        ^
  symbol:   method requestPermissions(String[],int)
  location: class Loader
src\org\renderdoc\renderdoccmd\Loader.java:17: error: package android.Manifest does not exist
        while(checkSelfPermission(android.Manifest.permission.WRITE_EXTERNAL_STORAGE)
                                                  ^
src\org\renderdoc\renderdoccmd\Loader.java:18: error: package android.content.pm does not exist
            != android.content.pm.PackageManager.PERMISSION_GRANTED)
                                 ^
9 errors
```

From what I can guess, the `javac` line specifies android.jar as a classpath which should provide these so I'm not sure what's going wrong. I'm not really sure to go about diagnosing it - any ideas?

cnorthrop commented:
Try changing the colon (`:`) to a semicolon (`;`) on the javac line, and maybe drop the path on obj (`./obj` -> `obj`)

baldurk commented:
Yeh it's the colon/semicolon thing. Actually getting a semicolon through is fiddly - without cmake either interpreting it as a list separator and replacing it with a space, or passing it all the way through to the command line without escaping it, breaking the command. I ended up doing:

```
    set(CLASS_PATH "${ANDROID_JAR}\;obj")
...
    COMMAND ${JAVA_BIN}/javac -d ./obj -source 1.7 -target 1.7 -bootclasspath ${RT_JAR} -classpath "${CLASS_PATH}" -sourcepath src src/org/renderdoc/renderdoccmd/*.java
```

which may also have the benefit of making it easier to vary the classpath by host platform.

cnorthrop commented:
I added a commit that should work cross platform.  If that works for you, we can squish it into the primary commit.

baldurk commented:
I tested it just now - the problem is `if(UNIX)` passes since it's checking the _target_ platform, which is android and so falls under UNIX. If you reverse the if/else cases and use `if(CMAKE_HOST_WIN32)` then it works OK.

FYI the `\` in `"${ANDROID_JAR}\;obj"` is used to escape the `;`, not as a windows path specifier, so I don't think you need to use a `/` on other platforms if it wasn't there already. Up to you though as I didn't test building on linux.

cnorthrop commented:
Heh, I guess I butchered that attempt.  Updated the commit to use host platform and remove the extra escape character.  Retested on Ubuntu.

baldurk commented:
Manually merged to v1.x and v0.x, thanks!
